File: perf/benchmarks/micro/hashmap_ops/hashmap_ops.cpp
// HashMap insert + lookup — stdlib hash-table micro-benchmark (C++).
#include <chrono>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>
#include <unordered_map>
#include <vector>

int main(int argc, char** argv) {
    int64_t n = (argc > 1) ? std::atoll(argv[1]) : 500'000LL;

    // Pre-build keys.
    std::vector<std::string> keys;
    keys.reserve(static_cast<size_t>(n));
    char buf[32];
    for (int64_t i = 0; i < n; ++i) {
        std::snprintf(buf, sizeof(buf), "key-%lld", static_cast<long long>(i));
        keys.emplace_back(buf);
    }

    auto t0 = std::chrono::steady_clock::now();
    std::unordered_map<std::string, int64_t> m;
    m.reserve(static_cast<size_t>(n));
    for (int64_t i = 0; i < n; ++i) {
        m.emplace(keys[static_cast<size_t>(i)], i);
    }
    int64_t cs = 0;
    for (int64_t i = 0; i < n; ++i) {
        cs ^= m[keys[static_cast<size_t>(i)]];
    }
    auto t1 = std::chrono::steady_clock::now();
    double ms = std::chrono::duration<double, std::milli>(t1 - t0).count();

    std::printf("CHECKSUM:%lld\n", static_cast<long long>(cs));
    std::printf("ELAPSED_MS:%.6f\n", ms);
    return 0;
}

File: perf/benchmarks/micro/hashmap_ops/hashmap_ops.py
# HashMap insert + lookup — stdlib hash-table micro-benchmark.
#
# Generates N synthetic string keys, inserts (key, i) pairs, then looks up
# each key once and accumulates a checksum from the values.
# Argument: N (default 500_000).

import sys
import time


def main() -> None:
    n = int(sys.argv[1]) if len(sys.argv) > 1 else 500_000
    # Pre-build keys outside the timed region so we measure the hash-map work.
    keys = [f"key-{i}" for i in range(n)]

    t0 = time.perf_counter_ns()
    m: dict[str, int] = {}
    for i, k in enumerate(keys):
        m[k] = i
    cs = 0
    for k in keys:
        cs ^= m[k]
    elapsed_ns = time.perf_counter_ns() - t0

    print(f"CHECKSUM:{cs}")
    print(f"ELAPSED_MS:{elapsed_ns / 1_000_000:.6f}")


if __name__ == "__main__":
    main()
